core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -951,14 +957,104 @@ wasi_path_remove_directory(wasm_exec_env_t exec_env, wasi_fd_t fd,
     return wasmtime_ssp_path_remove_directory(curfds, fd, path, path_len);
 }
 
+static __wasi_timestamp_t
+get_timeout_for_poll_oneoff(const wasi_subscription_t *in,
+                            uint32 nsubscriptions)
+{
+    __wasi_timestamp_t timeout = 0;
+    const __wasi_subscription_t *clock_subscription = NULL;
+
+    for (size_t i = 0; i < nsubscriptions; ++i) {
+        const __wasi_subscription_t *s = &in[i];
+        if (s->u.type == __WASI_EVENTTYPE_CLOCK
+            && (s->u.u.clock.flags & __WASI_SUBSCRIPTION_CLOCK_ABSTIME) == 0) {
+            clock_subscription = s;
+            break;
```

Comment:
@loganek / @wenyongh does it make sense to return the largest timeout amongst all clock subscriptions for final poll_oneoff timeout?

Reply:
It is really an issue, the timeout of each subscription may be different, not sure whether `poll_oneoff` will always return when the smallest timeout of all subscriptions is elapsed or not? If yes, I think here we should use the smallest timeout of all subscriptions.

It not, the algorithm is complex, here we should use the largest timeout, and record each subscription's left timeout (its original timeout - elapsed time), update them after `poll_oneoff` (L1030), and if the left timeout is < 0, remove it from the subscription list.

Seems the first one is reasonable, had better have a test for poll_oneoff under Linux to get the detail.

Reply:
iiuc, if there are multiple clock subscriptions and file descriptor (fd), [poll](https://linux.die.net/man/2/poll) waits until min timeout of all clock subscription has elapsed or when the fd is ready for I/O, whichever happens first. so I guess, it makes sense to take minima of all timeouts and return from this method (and use it in our while loop for timeout chaining).